Set origin incomplete on routes from MLAG iBGP peer

emilotep opened the pull request:
<!-- Bot will manage tag, ownership and code review, you should not update these fields-->
## Change Summary
Add variable knob and functionality for setting origin to incomplete on routes received from MLAG iBGP peer on l3leafs. The aim is to avoid suboptimal forwarding over the peerlink for some traffic.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [x] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)
Fixes #572 

## Component(s) name
eos_l3ls_evpn/README.md
eos_l3ls_evpn/templates/evpn-fabric-l3leaf-yml.j2
eos_l3ls_evpn/templates/fabric/bgp_base/leaf-router-bgp-peer-groups.j2
eos_l3ls_evpn/templates/fabric/bgp_base/route-maps.j2
eos_cli_config_gen/templates/eos/router-bgp.j2


eos_cli_config_gen
eos_l3ls_evpn

## Proposed changes
Add a new knob "mlag_ibgp_origin_incomplete: true/false, default -> true" under fabric variables, either as default for all l3leafs or individually per l3leaf node group to override the default.

Additions to eos_l3ls_evpn and eos_cli_config_gen templates have been made to define a route-map RM-MLAG-PEER that sets the origin and applies it inbound to the MLAG iBGP peer group: MLAG-IPv4-UNDERLAY-PEER. The route-map is not created if the knob is set to false.

The new variable is documented in the README for eos_l3ls_evpn role under the section that deals with fabric variables.

### Example:

    l3leaf:
      defaults:
        ...
        mlag_ibgp_origin_incomplete: true
      node_groups:
        leaf1:
          ...
          mlag_ibgp_origin_incomplete: false

## How to test
Have tried rendering EVPN configurations with most combinations of defining the knob under l3leaf defaults as well as overriding the default per node group. Templates render as expected regardless of whether the knob is undefined (implicitly true) or defined (explicitly true or false) at different levels.

## Checklist:
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

emilotep committed:
Set origin incomplete on routes from MLAG iBGP peer

Add variable knob and functionality for setting origin to incomplete
on routes received from MLAG iBGP peer on l3leafs. The aim is to avoid
suboptimal forwarding over peerlink for some traffic.

The new knob is "mlag_ibgp_origin_incomplete: true/false" and is
defined under fabric variables, either as default for all l3leafs or
individually per l3leaf node group. The setting defaults to true if
omitted.

Additions to eos_l3ls_evpn and eos_cli_config_gen have been made to
define a route-map RM-MLAG-PEER that sets the origin and applies it to
the MLAG iBGP peer group: MLAG-IPv4-UNDERLAY-PEER.

The knob is documented in the README for eos_l3ls_evpn role under the
section that deals with fabric variables.

emilotep committed:
Update CI artifacts

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/README.md:
```diff
@@ -584,6 +584,9 @@ l3leaf:
       mlag_interfaces: [ < ethernet_interface_3 >, < ethernet_interface_4 >]
 
+      # Set origin of routes received from MLAG iBGP peer to incomplete
+      mlag_ibgp_origin_incomplete: < true | false -> default true >
```
Are we sure we want to make this the new default?

emilotep replied:
I guess it's a question of whether it can break any current environment, and also whether it should be considered best practice to use it.

ClausHolbechArista replied:
Let us add a good description alongside this knob. To make it clear why we have this.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/fabric/bgp_base/route-maps.j2:
```diff
@@ -38,3 +38,16 @@
+          - "origin incomplete"
+{%         endif %}
+{%     endif %}
+{% endif %}
```
missing newline at the end of the file

emilotep replied:
Fixed in commit a47a8fb.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/fabric/bgp_base/leaf-router-bgp-peer-groups.j2:
```diff
@@ -17,6 +17,9 @@
       maximum_routes: 12000
       send_community: true
+{%             if leaf.mlag_ibgp_origin_incomplete == true %}
+      route_map_in: RM-MLAG-PEER
```
I would suggest calling this `RM-MLAG-PEER-IN`

emilotep replied:
Fixed in commit a47a8fb.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/fabric/bgp_base/route-maps.j2:
```diff
@@ -38,3 +38,16 @@
+{% if type == "l3leaf" %}
+{%     if leaf.mlag == true %}
+{%         if leaf.mlag_ibgp_origin_incomplete == true %}
+  RM-MLAG-PEER:
```
I would suggest calling this RM-MLAG-PEER-IN

emilotep replied:
Fixed in commit a47a8fb.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:
```diff
@@ -59,6 +59,9 @@ router bgp {{ router_bgp.as }}
 {%             if router_bgp.peer_groups[peer_group].timers is defined and router_bgp.peer_groups[peer_group].timers is not none %}
    neighbor {{ peer_group }} timers {{ router_bgp.peer_groups[peer_group].timers }}
 {%             endif %}
+{%             if router_bgp.peer_groups[peer_group].route_map_in is defined and router_bgp.peer_groups[peer_group].route_map_in is not none %}
```
please refactor to use the new `is arista.avd.defined` test

emilotep replied:
Fixed in commit a47a8fb.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-l3leaf-yml.j2:
```diff
@@ -81,6 +81,13 @@
 {%             else %}
 {%                 set switch.mlag = l3leaf.defaults.mlag | default(true) %}
 {%             endif %}
+{%             if l3leaf.node_groups[l3leaf_node_group].mlag_ibgp_origin_incomplete is defined and l3leaf.node_groups[l3leaf_node_group].mlag_ibgp_origin_incomplete is not none %}
```
please refactor to use the new `| arista.avd.default` filter

emilotep replied:
Fixed in commit 49f51f6.

ClausHolbechArista reviewed (changes requested):
A few changed and also we need to discuss setting this as default. Reach out if you need help on the new test/filter.

emilotep committed:
Changes requested in PR #600

- Add newline at end of route-maps.j2
- Renaming new route-map from RM-MLAG-PEER to RM-MLAG-PEER-IN
- Refactoring to use new arista.avd.defined test in /eos_cli_config_gen/templates/eos/router-bgp.j2

emilotep committed:
Changes in evpn-fabric-l3leaf-yml.j2 template

Refactored to using arista.avd.default filter for added variable.

ClausHolbechArista commented on Makefile:
```diff
@@ -2,7 +2,7 @@ CURRENT_DIR = $(shell pwd)
 ANSIBLE_TEST ?= $(shell which ansible-test)
 # option to run ansible-test sanity: must be either venv or docker (default is docker)
-ANSIBLE_TEST_MODE ?= docker
+ANSIBLE_TEST_MODE ?= venv
```
is this on purpose to change this as part of this PR?

titom73 replied:
It should be covered in another PR as we have to validate there is no blocker with CI runner using this molecule provider

emilotep replied:
This was unintentional. Changed back to docker in commit 0b05dd9.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/route-maps.yml:
```diff
@@ -32,3 +32,9 @@ route_maps:
         type: permit
         set:
           - as-path match all replacement auto
+  RM-MLAG-PEER:
```
RM-MLAG-PEER-**IN** :)

emilotep replied:
Fixed in commit 0b05dd9

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/router-bgp-v4-evpn.yml:
```diff
@@ -31,6 +31,7 @@ router_bgp:
       password: "vnEaG8gMeQf3d3cN6PktXQ=="
       maximum_routes: 12000
       send_community: true
+      route_map_in: RM-MLAG-PEER
```
RM-MLAG-PEER-**IN** :)

emilotep replied:
Fixed in commit 0b05dd9.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/molecule.yml:
```diff
@@ -14,7 +14,7 @@ scenario:
   name: galaxy
 driver:
-  name: docker
+  name: delegated
```
is this on purpose to change this as part of this PR?

titom73 replied:
It should be covered in another PR as we have to validate there is no blocker with CI runner using this molecule provider

emilotep replied:
This was unintentional. Changed back to name: docker in commit 0b05dd9.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:
```diff
@@ -59,6 +59,9 @@ router bgp {{ router_bgp.as }}
 {%             endif %}
+{%             if router_bgp.peer_groups[peer_group].route_map_in is arista.avd.defined %}
+   neighbor {{ peer_group }} route-map {{ router_bgp.peer_groups[peer_group].route_map_in }} in
+{%             endif %}
```
please add support for route_map_out while you are at it :)

ClausHolbechArista replied:
Also add both to the readme.md for eos_cli_config_gen

emilotep replied:
Fixed in commit 29761e3.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/fabric/bgp_base/route-maps.j2:
```diff
@@ -38,3 +38,16 @@
+{%     if leaf.mlag == true %}
+{%         if leaf.mlag_ibgp_origin_incomplete == true %}
+  RM-MLAG-PEER-IN:
+    sequence_numbers:
```
Please add a good description on the route-map to make it clear what the purpose is. It is not a very intuitive workaround.

emilotep replied:
Config description added in commit 6c952bb. Documentation will follow.

emilotep replied:
Documentation updated in commit 0b05dd9.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/fabric/bgp_base/route-maps.j2:
```diff
@@ -38,3 +38,16 @@
+{# Origin Incomplete for MLAG iBGP learned routes #}
+{% if type == "l3leaf" %}
+{%     if leaf.mlag == true %}
+{%         if leaf.mlag_ibgp_origin_incomplete == true %}
```
Add another condition to check if we use bgp for underlay (look at other bgp templates for examples)

emilotep replied:
Fixed in commit 0b05dd9.

emilotep committed:
Add description to RM-MLAG-PEER-IN

Add descriptive text to the new route-map to clarify the
purpose of the added configuration.

emilotep committed:
Improve documentation, correcting molecule/Makefiles

- Undid unintentional changes to Makefile and molecule.yml
- Update molecule files with correct route-map name
- Added condition in route-maps.yml to check for underlay proto = ebgp
- Improved documentation of new variable knob

ClausHolbechArista commented:
@emilotep So after the latest changes you need to rerun molecule if you can. And then you have a merge conflict for eos_l3ls_evpn/README.md, so I suggest to do a rebase.

emilotep committed:
Update CI artifacts

emilotep committed:
Update CI artifacts

emilotep committed:
Add support for route_map_out in eos_cli_config_gen

- Support for route_map_out added in router-bgp.j2 template
- Added eos_cli_config_gen role documentation for route_map_in and _out
- Added route_map_out to molecule
- Reran molecule and linting

titom73 reviewed (approved):
Looks good to me!
